README.md
repostatus.org
==============

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)

A standard to easily communicate to humans and machines the development/support and usability status of software repositories/projects.

For the majority of documentation and human-readable text, see https://www.repostatus.org/ or the [gh-pages branch](https://github.com/jantman/repostatus.org/tree/gh-pages) from which it is built.

Please feel free to leave comments as Issues, or open pull requests.

Community Involvement
---------------------

This project seems to have gained a lot more interest than I thought it would. As of April, 2017 there are [over 1,200 references on GitHub](https://github.com/search?l=&q=http%3A%2F%2Fwww.repostatus.org%2Fbadges%2F+-user%3A%22jantman%22&ref=advsearch&type=Code&utf8=%E2%9C%93)
to repostatus.org badge URLs. I do *not* want to be the sole person making decisions for this project. I encourage everyone who finds
it useful to watch [the repo on GitHub](https://github.com/jantman/repostatus.org) and provide their feedback in discussions, especially the
issues with the [discussion](https://github.com/jantman/repostatus.org/issues?q=is%3Aopen+is%3Aissue+label%3Adiscussion) or
["needs decision"](https://github.com/jantman/repostatus.org/issues?q=is%3Aopen+is%3Aissue+label%3Adiscussion+label%3A%22needs+decision%22)
labels. I'm handling the code updates, but I very much want this project to be driven based on consensus of those who use it.

Contributing
------------

For changes to the site, text, or anything other than the badges themselves (and their descriptions and sample markup),
simply cut a pull request against the master branch. The content that appears on the website (in the gh-pages branch)
comes from ``gh_pages/`` in master. Note that some of it (described below) is generated programmatically.

The badges (SVG), their descriptions and their sample markup are generated by a [Fabfile](http://www.fabfile.org/). If you're looking
to add a new badge or make changes to an existing one, simply update the ``badge_info`` dictionary at the top of ``fabfile.py`` and
then run ``fab make_badges`` (requires Python and some packages; see the comment at the top of the file for requirements). This will
regenerate all badges, metadata and samples into ``badges/latest``. You can then cut a pull request for this; a version number will
be assigned at merge time. Please remember to also update ``gh_pages/index.md`` for any badge changes.

Release Process
---------------

1. Get everything included in the release merged into master.
2. Assign a version number. In general, patch versions should only be assigned for releases that fix trivial (i.e. spelling)
issues or touch things other than the badges and JSON (i.e. the markup samples). Minor versions should be assigned to
changes that correct grammatical or spelling errors, or graphical elements. Major versions must be assigned to any changes
that add or remove badges, or alter the meaning of existing badges.
3. Re-run ``fab make_badges`` and ensure there are no new changes.
4. Run ``fab version_badges:x.y.z`` (where ``x.y.z`` is the version number).
5. Add a ``CHANGELOG.md`` entry.
6. Run ``fab badges2pages`` to copy the badges under ``gh-pages/``
6. Run ``fab publish`` to push changes to the gh-pages branch.
7. Review the diff of gh-pages against origin.
8. Assuing all is well, push gh-pages to origin. The changes are now live.
9. Tag master with the version number (use GitHub Releases)


history.md
---
layout: minimal
---

## History

I have about 50 repositories on GitHub, in varying states of completion and support. Some of them are totally abandoned, some are active (with users and everything!) and some
are just repositories holding a ``README.md`` describing some idea I had. I was about to embark on a new project, and realized I needed to get a handle on what I currently have
first. After some thought, I realized that an awful lot of people are in the same position. And even more so, I can't count the number of times I've dug through a repository –
and even emailed the author – to try and figure out if it was still active or not.

So, I came up with this idea. I had to stop myself before I over-engineered it too much with a formal specification and versioned API and everything.

I originally thought that this should be a feature built-in to GitHub. But for anyone who's seen [isaacs/github](https://github.com/isaacs/github), you know how unlikely
that is, and how ironically secretive GH is about feature requests and fixing bugs, and communication with the community in general. So I [opened an issue](https://github.com/isaacs/github/issues/312)
anyway, but I doubt it will go anywhere. Then I started thinking about how to do this without GitHub involved, which would give the added bonus that it's hosting-service
agnostic (and VCS-agnostic too, so it'll work for that crowd of Mercurial users too).

The blog post where I first laid this out is [here](https://blog.jasonantman.com/2014/12/idea-for-a-generic-method-to-communicate-repositoryproject-status/). Sure, this is a tiny
little thing that is more of a community suggestion than a standard, but it'd sure be nice if we could all easily tell each other what's done and what isn't, and more importantly,
what we want to spend time on again "one day."


parsers/check_github_repos.py
#!/usr/bin/env python
"""
This script uses the GitHub API (via PyGithub) to list all of a user's
repositories (optionally including forks), searches each repository for
RepoStatus information in ``repostatus.org``, ``.repostatus.org``, or any file
in the root of the repository starting with ``readme`` (case-insensitive),
and outputs a listing of repos and their statuses.

============
Requirements
============

* Python 2.7 or newer.
* pygithub - `pip install pygithub` <https://github.com/jacquev6/PyGithub>
* requests - `pip install requests`

You'll need to set your GitHub API token in your git config;
use `git config --global github.token <your token>` to set it
if not already present.

=========
Copyright
=========

Copyright 2014-2018 Jason Antman <jason@jasonantman.com> <https://jantman.github.io/>
Free for any use provided that patches are submitted back to me.

The latest version of this script can be found at:
https://github.com/jantman/repostatus.org/blob/master/parsers/repostatusorg_list_repo_status.py

=========
CHANGELOG
=========

2018-04-01 jantman:
- Give user some help if non-standard library modules can't be imported
- Make RepoStatusOrg_GitHub_Checker a new-style class
- Update summary at top of this docstring
- Add -F/--fail-on-unknown option
- Python3 fixes

2016-05-18 jantman:
- add links to repo in HTML output

2016-05-17 jantman:
- add JSON and HTML output options

2014-12-25 jantman:
- initial script
"""

import sys
import argparse
import logging
import re
import subprocess
from base64 import b64decode
import json
from datetime import datetime

try:
    import requests
except ImportError:
    sys.stderr.write(
        'ERROR importing "requests". If it is not installed, please '
        '"pip install requests"\n'
    )
    raise
try:
    from github import Github
except ImportError:
    sys.stderr.write(
        'ERROR importing "github". If it is not installed, please '
        '"pip install pygithub"\n'
    )
    raise

FORMAT = "[%(levelname)s %(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"
logging.basicConfig(level=logging.INFO, format=FORMAT)


class RepoStatusOrg_GitHub_Checker(object):
    """ check a user's GitHub repos for repostatus.org status identifiers """

    readme_re = re.compile(r'^readme.*$', flags=re.I)
    url_re = re.compile(r'http[s]?:\/\/.*repostatus\.org\/badges\/(.+)\/(.+)\.svg', flags=re.I)

    def __init__(self, verbose=False):
        self.logger = logging.getLogger(self.__class__.__name__)
        if verbose:
            self.logger.setLevel(logging.DEBUG)
        # try to get GitHub credentials
        try:
            token = subprocess.check_output(['git', 'config', '--global', 'github.token']).strip()
            if isinstance(token, type(b'')):
                token = token.decode()
            self.logger.debug("got github token: {t}".format(t=token))
        except subprocess.CalledProcessError:
            self.logger.error("ERROR: no github token found. Set 'git config --global github.token' to your API token.")
            raise SystemExit(1)
        self.logger.debug("connecting to GitHub API")
        self.g = Github(login_or_token=token)

    def check(self, github_user, include_forks=False):
        """
        Check all repositories of a given GitHub user (or organization) for
        repostatus.org identifiers.

        returns a dictionary of repo name to status name (or None if no status found)

        :param github_user: github user or organization to check repos for, or None for logged in user
        :type github_user: string
        :rtype: dict
        """
        res = {}
        if github_user is None:
            github_user = self.g.get_user().login
        self.username = github_user
        self.logger.debug("checking repos for user {u}".format(u=github_user))
        user = self.g.get_user(github_user)
        if user.type == 'Organization':
            self.logger.debug("user is an Organization; using organization instead")
            user = self.g.get_organization(user.login)
        self.logger.debug("user has {r} public repos and {p} owned private repos".format(r=user.public_repos, p=user.owned_private_repos))
        repos = user.get_repos()
        count = 0
        forks = 0
        for repo in repos:
            if repo.fork and not include_forks:
                self.logger.debug("ignoring fork: {r}".format(r=repo.name))
                forks += 1
                continue
            count += 1
            self.logger.debug("checking repo {r}".format(r=repo.name))
            candidates = self._find_candidate_files(repo)
            self.logger.debug("found {c} candidate files".format(c=len(candidates)))
            if len(candidates) == 0:
                continue
            status = self._find_status_for_files(repo, candidates)
            if status is not None:
                self.logger.debug("found status {s} for repo {r}".format(s=status, r=repo.name))
                res[repo.name] = status
            else:
                self.logger.debug("found no status for repo {r}".format(r=repo.name))
                res[repo.name] = None
        self.logger.debug("checked {c} repos for user; ignored {f} forks".format(c=count, f=forks))
        return res

    def _find_status_for_files(self, repo, flist):
        """
        Given a list of files to search, returns the repostatus.org version
        and status name of the first matching status identifier URL found;
        searches the files in list order. Returns None if no match found

        :param repo: repository to check
        :type repo: github.Repository.Repository
        :param flist: list of files to search through, in order
        :type flist: list of strings (file paths)
        :rtype: 2-tuple (version, status name) or None
        """
        for f in flist:
            content = repo.get_file_contents(f)
            s = ''
            if content.encoding == 'base64':
                s = b64decode(content.content)
                if isinstance(s, type(b'')):
                    s = s.decode()
            else:
                self.logger.error(
                    "unknown encoding '%s' on file %s in repository %s",
                    content.encoding, content.path, repo.name
                )
            res = self.url_re.search(s)
            if res is not None:
                self.logger.debug("Match found in {f}: {u}".format(f=content.path, u=res.group(0)))
                return (res.group(1), res.group(2))
        return None

    def _find_candidate_files(self, repo):
        """
        Return a list of all files in the top directory/path of the repository
        which should be examined for a repostatus identifier.
        List is in the order they should be checked.

        :param repo: repository to check
        :type repo: github.Repository.Repository
        :rtype: list of string filenames
        """
        files = []
        for x in repo.get_dir_contents('/'):
            if x.type != 'file':
                continue
            files.append(x.name)
        candidates = []
        # sort files lexicographically
        for fname in sorted(files, key=lambda x: x.lower()):
            if self.readme_re.match(fname):
                candidates.append(fname)
        if '.repostatus.org' in files:
            candidates.append('.repostatus.org')
        if 'repostatus.org' in files:
            candidates.append('repostatus.org')
        return candidates


def parse_args(argv):
    """
    parse command line arguments/options
    """
    p = argparse.ArgumentParser(description='repostatus.org GitHub parser')
    p.add_argument('-v', '--verbose', dest='verbose', action='store_true', default=False,
                   help='verbose output (internal debugging).')
    p.add_argument('-u', '--user', dest='user', type=str, default=None,
                   help='GitHub user or organization to check repos for; defaults to current user')
    p.add_argument('-f', '--forks', dest='forks', action='store_true', default=False,
                   help='also include forks')
    p.add_argument('-o', '--output-format', dest='format', action='store',
                   choices=['text', 'json', 'html'],
                   default='text', help='output format - (text|json|html) - default "text"')
    p.add_argument('-F', '--fail-on-unknown', dest='fail_on_unknown',
                   action='store_true', default=False,
                   help='exit 1 if any repos have an unknown status')
    args = p.parse_args(argv)
    return args


def htmlout(output, username):
    out = """
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="https://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=windows-1252">
    <title>repostatus.org parse results for {user}</title>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.2.3/jquery.min.js" type="text/javascript"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.26.1/js/jquery.tablesorter.js" type="text/javascript"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.26.1/js/jquery.tablesorter.widgets.js" type="text/javascript"></script>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jquery.tablesorter/2.26.1/css/theme.default.min.css" />
  </head>
  <body>
    <table id="myTable" class="tablesorter">
      <thead>
        <th>Repo Name</th>
        <th>Status</th>
      </thead>
      <tbody>
{tbody}
      </tbody>
    </table>
    <p><em>Generated by repostatus.org check_github_repos.py parser at {dt}</em></p>
{script}
  </body>
</html>
"""
    script = """
    <script type="text/javascript">
      $(document).ready(function()
        {
          $("#myTable").tablesorter();
        }
      );
    </script>
"""
    tbody = ''
    for repo in sorted(output):
        href = 'https://github.com/%s/%s' % (username, repo)
        tbody += "        <tr><td><a href=\"%s\">%s</a></td><td>%s</td></tr>\n" % (
            href, repo, output[repo]
        )
    dt = datetime.now().isoformat()
    out = out.format(user=username, dt=dt, tbody=tbody, script=script)
    return out


if __name__ == "__main__":
    args = parse_args(sys.argv[1:])
    # initialize the class
    checker = RepoStatusOrg_GitHub_Checker(verbose=args.verbose)
    # run the check
    results = checker.check(args.user, include_forks=args.forks)
    total = 0
    unknown = []
    output = {}
    maxlen = 0
    for repo in results:
        if results[repo] is None:
            s = 'UNKNOWN'
            unknown.append(repo)
        else:
            s = results[repo][1]
        total += 1
        output[repo] = s
        if len(repo) > maxlen:
            maxlen = len(repo)
    if args.format == 'html':
        print(htmlout(output, checker.username))
    elif args.format == 'json':
        print(json.dumps(output))
    else:
        # text
        fs = '{:<%d}   {}' % ( maxlen + 1 )
        for repo in sorted(output):
            print(fs.format(repo, output[repo]))
    checker.logger.info(
        "Found %d repos, %d with unknown status", total, len(unknown)
    )
    if len(unknown) > 0:
        checker.logger.info('Unknown repos: %s', unknown)
        if args.fail_on_unknown:
            raise SystemExit(1)


parsers/github_simple_checker.py
#!/usr/bin/env python
"""
This is a simple repostatus.org status parser for GitHub repositories.

It is **NOT** compliant with the official spec; this is just something
I use for my own repos, when I know beyond a doubt that the status
badge is in a README file that is rendered properly and automatically
by GitHub.

This requires the requests package, which can be installed via
`pip install requests`.

"""

import sys
import optparse
import requests
import re

def get_repostatusorg_for_url(url):
    """ check and return repo status """
    r = requests.get(url)
    if r.status_code != 200:
        raise SystemExit("ERROR: Got HTTP status code {c} for url: {u}".format(c=r.status_code, u=url))
    m = re.search(r'http[s]?:\/\/.*repostatus\.org\/badges\/(.+)\/(.+)\.svg', r.text, flags=re.I)
    if m is None:
        # no match found
        return None
    # ok, the status is in capture group 2 from the above search
    return m.group(2)

def parse_args(argv):
    """ parse arguments/options """
    p = optparse.OptionParser(usage="USAGE: github_simple_checker.py -g <username/reponame> OR -u <url>")
    p.add_option('-g', '--github', action='store', dest='github', type=str, help="github URL portion, in <username>/<reponame> format")
    p.add_option('-u', '--url', action='store', dest='url', type=str, help='URL to check for repostatus badge (for non-GitHub sites)')
    options, args = p.parse_args(argv)

    if options.github and options.url:
        raise SystemExit("ERROR: you must specify either -g/--github or -u/--url not both.")
    
    return options

if __name__ == "__main__":
    # parse and validate options/arguments
    opts = parse_args(sys.argv[1:])
    if opts.github:
        # if github, just build a URL out of the string we were given
        url = 'https://github.com/' + opts.github
    elif opts.url:
        url = opts.url
    else:
        raise SystemExit("ERROR: you must specify either -g/--github or -u/--url; see --help for more info.")
    # get the status
    status = get_repostatusorg_for_url(url)
    if status is None:
        status = 'not specified'
    print("{url} status: {status}".format(url=url, status=status))


parsers/repostatusorg_list_repo_status.py
#!/usr/bin/env python
"""
This is an example parser for the <https://www.repostatus.org/> specification.
It checks a single project/repository on disk, or a directory containing multiple
projects, and reports the status(es).

NOTE: This script works with Python 2.7 or newer.

=============
Example Usage
=============

On a single directory, containing a README.md file with the identifier string present in it:

    $ repostatusorg_list_repo_status.py -p ~/GIT/repostatus.org/
    /home/jantman/GIT/repostatus.org/: active

In pwd, containing a README.md file with the identifier string present:

    $ repostatusorg_list_repo_status.py
    ./: active

On a directory containing multiple git clones, some of which have the identifier present:

    $ parsers/repostatusorg_list_repo_status.py -p ~/GIT
    /home/jantman/GIT/php-nagios-xml: unsupported
    /home/jantman/GIT/ec2machines: active
    /home/jantman/GIT/nodejs-rpm-centos5: inactive
    /home/jantman/GIT/updatewatcher: concept
    /home/jantman/GIT/repostatus.org: active
    /home/jantman/GIT/TuxTruck-wxPython: abandoned

=========
Copyright
=========

Copyright 2014 Jason Antman <jason@jasonantman.com> <http://www.jasonantman.com>
Free for any use provided that patches are submitted back to me.

The latest version of this script can be found at:
https://github.com/jantman/repostatus.org/blob/master/parsers/repostatusorg_list_repo_status.py

=========
CHANGELOG
=========

2014-12-25 jantman:
- initial script
"""

import sys
import argparse
import logging
import os
import re

FORMAT = "[%(levelname)s %(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"
logging.basicConfig(level=logging.ERROR, format=FORMAT)


class RepoStatusOrg_Checker:
    """ check directories on disk for repostatus.org status identifier """

    readme_re = re.compile(r'^readme.*$', flags=re.I)
    url_re = re.compile(r'http[s]?:\/\/.*repostatus\.org\/badges\/(.+)\/(.+)\.svg', flags=re.I)
    
    def __init__(self, verbose=False):
        self.logger = logging.getLogger(self.__class__.__name__)
        if verbose:
            self.logger.setLevel(logging.DEBUG)

    def check(self, path):
        """
        check a path, that may be a single project
        or a directory containing multiple projects

        returns a dictionary of path to status name (or None if no status found)

        :param path: path to check
        :type path: string
        :rtype: dict
        """
        # first, see if we have a matching file in this path
        candidates = self._find_candidate_files(path)
        self.logger.debug("Found {c} candidate files in path {p}: {candidates}".format(c=len(candidates),
                                                                                       p=path,
                                                                                       candidates=candidates))
        if len(candidates) > 0:
            # this path is a project itself, check for a match and return that
            self.logger.debug("Found candidate files in {p}; checking only this path".format(p=path))
            res = self._find_status_for_files(candidates)
            return {path: res}
        # else we found no candidates; look for child directories that might be projects
        self.logger.debug("Found no candidate files in {p}; checking subdirectories".format(p=path))
        res = {}
        for d in os.listdir(path):
            dpath = os.path.join(path, d)
            if not os.path.isdir(dpath) or d == '.' or d == '..':
                self.logger.debug("skipping non-directory or system directory: {dpath}".format(dpath=dpath))
                continue
            candidates = self._find_candidate_files(dpath)
            if len(candidates) < 1:
                self.logger.debug("found 0 candidates in directory: {dpath}".format(dpath=dpath))
                continue
            self.logger.debug("Found candidate files in subdirectory: {d}".format(d=dpath))
            res[dpath] = self._find_status_for_files(candidates)
        return res

    def _find_status_for_files(self, flist):
        """
        Given a list of files to search, returns the repostatus.org version
        and status name of the first matching status identifier URL found;
        searches the files in list order. Returns None if no match found

        :param flist: list of files to search through, in order
        :type flist: list of strings (file paths)
        :rtype: 2-tuple (version, status name) or None
        """
        for f in flist:
            with open(f, 'r') as fh:
                content = fh.read()
            res = self.url_re.search(content)
            if res is not None:
                self.logger.debug("Match found in {f}: {u}".format(f=f, u=res.group(0)))
                return (res.group(1), res.group(2))
        return None
        
    def _find_candidate_files(self, path):
        """
        Return a list of all files under a given path,
        which should be examined for a repostatus identifier.
        List is in the order they should be checked.
        """
        candidates = []
        files = os.listdir(path)
        # sort files lexicographically
        for fname in sorted(files, lambda x,y: cmp(x.lower(), y.lower()) or cmp(x,y)):
            if self.readme_re.match(fname) and os.path.isfile(os.path.join(path, fname)):
                candidates.append(os.path.join(path, fname))
        for fname in [os.path.join(path, '.repostatus.org'), os.path.join(path, 'repostatus.org')]:
            if os.path.exists(fname) and os.path.isfile(fname):
                candidates.append(fname)
        return candidates
            

def parse_args(argv):
    """
    parse command line arguments/options
    """
    p = argparse.ArgumentParser(description='Sample python script skeleton.')
    p.add_argument('-v', '--verbose', dest='verbose', action='store_true', default=False,
                   help='verbose output (internal debugging).')
    p.add_argument('-p', '--path', dest='path', type=str, default=os.getcwd(),
                   help='path to a project or directory of projects to check; default is cwd')
    args = p.parse_args(argv)
    return args

if __name__ == "__main__":
    args = parse_args(sys.argv[1:])
    # initialize the class
    checker = RepoStatusOrg_Checker(verbose=args.verbose)
    # run the check
    result = checker.check(args.path)
    for dir_path in result:
        status = 'unknown'
        if result[dir_path] is not None:
            status = result[dir_path][1]
        print("{dirpath}: {status}".format(dirpath=dir_path, status=status))


fabfile.py
# -*- coding: utf-8 -*-
"""
Repostatus.org fabfile – this is used to build badges and push to GitHub Pages.

Requires Python (only tested with 2.7), Fabric and ghp-import.

requirements (and tested versions):

Fabric==1.8.1
ghp-import==0.4.0
requests==2.7.0

"""

from fabric.api import local
import os
import re
import requests
import json
import shutil

badge_info = {
    'concept': {
        'shield_url': 'http://img.shields.io/badge/repo%20status-Concept-ffffff.svg',
        'description': "Minimal or no implementation has been done yet, or the repository is only intended to be a limited example, demo, or proof-of-concept.",
        'display_name': 'Concept',
    },
    'wip': {
        'shield_url': 'http://img.shields.io/badge/repo%20status-WIP-yellow.svg',
        'description': "Initial development is in progress, but there has not yet been a stable, usable release suitable for the public.",
        'display_name': 'WIP',
    },
    'suspended': {
        'shield_url': 'http://img.shields.io/badge/repo%20status-Suspended-orange.svg',
        'description': "Initial development has started, but there has not yet been a stable, usable release; work has been stopped for the time being but the author(s) intend on resuming work.",
        'display_name': 'Suspended',
    },
    'abandoned': {
        'shield_url': 'http://img.shields.io/badge/repo%20status-Abandoned-red.svg',
        'description': "Initial development has started, but there has not yet been a stable, usable release; the project has been abandoned and the author(s) do not intend on continuing development.",
        'display_name': 'Abandoned',
    },
    'active': {
        'shield_url': 'http://img.shields.io/badge/repo%20status-Active-brightgreen.svg',
        'description': "The project has reached a stable, usable state and is being actively developed.",
        'display_name': 'Active',
    },
    'inactive': {
        'shield_url': 'http://img.shields.io/badge/repo%20status-Inactive-yellowgreen.svg',
        'description': "The project has reached a stable, usable state but is no longer being actively developed; support/maintenance will be provided as time allows.",
        'display_name': 'Inactive',
    },
    'unsupported': {
        'shield_url': 'http://img.shields.io/badge/repo%20status-Unsupported-lightgrey.svg',
        'description': "The project has reached a stable, usable state but the author(s) have ceased all work on it. A new maintainer may be desired.",
        'display_name': 'Unsupported',
    },
    "moved": {
        'shield_url': 'http://img.shields.io/badge/repo%20status-Moved-red.svg',
        'description': 'The project has been moved to a new location, and the version at that location should be considered authoritative.',
        'display_name': 'Moved'
    }
}

def _download_media(url, fname):
    """ download the given binary URL to fname """
    r = requests.get(url, stream=True)
    if r.status_code != 200:
        raise SystemExit("%s returned status code %d" % (url, r.status_code))
    with open(fname, 'wb') as fh:
        for chunk in r.iter_content():
            fh.write(chunk)
        fh.flush()

def _make_badge_markup(badge_name, display_name, description, url, savedir):
    """ generate example markup for a badge, write to disk under savedir """
    if badge_name == 'moved':
        moved_to = 'http://example.com'
        moved = 'to %s ' % moved_to
    else:
        moved_to = None
        moved = ''
    alt = "Project Status: {statuscap} {moved}– {desc}".format(
        desc=description,
        statuscap=display_name,
        moved=moved
    )
    target = "https://www.repostatus.org/#{status}".format(status=badge_name)
    with open(os.path.join(savedir, '{n}_md.txt'.format(n=badge_name)), 'w') as fh:
        fh.write(_format_md(url, target, alt, moved_to))
    with open(os.path.join(savedir, '{n}_html.txt'.format(n=badge_name)), 'w') as fh:
        fh.write(_format_html(url, target, alt, moved_to))
    with open(os.path.join(savedir, '{n}_rst.txt'.format(n=badge_name)), 'w') as fh:
        fh.write(_format_rst(url, target, alt, moved_to))

def _format_md(url, target, alt, moved_to=None):
    if moved_to is None:
        moved = ''
    else:
        moved = ' to [%s](%s)' % (moved_to, moved_to)
    s = "[![{alt}]({url})]({target}){moved}\n".format(
        target=target,
        url=url,
        alt=alt,
        moved=moved
    )
    return s

def _format_rst(url, target, alt, moved_to=None):
    if moved_to is None:
        return '.. image:: {url}\n   :alt: {alt}\n   :target: {target}\n'.format(
            url=url,
            target=target,
            alt=alt
        )
    s = '|repostatus| to `%s <%s>`_\n\n' % (moved_to, moved_to)
    s += '.. |repostatus| image:: {url}\n   :alt: {alt}\n   ' \
        ':target: {target}\n'.format(
            url=url,
            target=target,
            alt=alt
        )
    return s

def _format_html(url, target, alt, moved_to=None):
    if moved_to is None:
        moved = ''
    else:
        moved = ' to <a href="%s">%s</a>' % (moved_to, moved_to)
    s = '<a href="{target}"><img src="{url}" alt="{alt}" /></a>{moved}\n'.format(
        url=url,
        target=target,
        alt=alt,
        moved=moved
    )
    return s

def make_badges():
    """ Regenerate the badges into badges/latest """
    if not os.path.exists('badges/latest'):
        os.makedirs('badges/latest')
    badge_data = {}
    for name in badge_info:
        badge_data[name] = {
            'description': badge_info[name]['description'],
            'url': 'https://www.repostatus.org/badges/latest/{name}.svg'.format(name=name),
            'display_name': badge_info[name]['display_name'],
        }
    with open('badges/latest/badges.json', 'w') as fh:
        fh.write(json.dumps(badge_data, indent=2, sort_keys=True))
    print("badge info written to badges/latest/badges.json")
    for name, _dict in badge_info.items():
        _download_media(_dict['shield_url'], 'badges/latest/{n}.svg'.format(n=name))
        _make_badge_markup(name, _dict['display_name'], _dict['description'], badge_data[name]['url'], 'badges/latest')
    print("badge images and markup written to badges/latest")

def version_badges(ver):
    """Copy the latest badges into a versioned directory; update related files"""
    if not re.match(r'\d+\.\d+\.\d+', ver):
        raise SystemExit("Error: %s does not appear to be an x.y.z semver version" % ver)
    badgedir = os.path.join('badges', ver)
    if os.path.exists(badgedir):
        raise SystemExit("Error: badge version %s already present!" % ver)
    # copy latest
    print("Copying badges/latest to badges/%s" % ver)
    shutil.copytree('badges/latest', 'badges/%s' % ver)
    # update URLs
    print("Updating URLs in badges/%s/*.txt" % ver)
    for fname in os.listdir(badgedir):
        if not fname.endswith('.txt'):
            continue
        fpath = os.path.join(badgedir, fname)
        with open(fpath, 'r') as fh:
            content = fh.read()
        content = content.replace('/badges/latest/', '/badges/%s/' % ver)
        with open(fpath, 'w') as fh:
            fh.write(content)
    print("Updating badges/%s/badges.json" % ver)
    with open('badges/%s/badges.json' % ver, 'r') as fh:
        content = fh.read()
    content = content.replace('/badges/latest/', '/badges/%s/' % ver)
    with open('badges/%s/badges.json' % ver, 'w') as fh:
        fh.write(content)

def badges2pages():
    """Copy badges/ to gh_pages/badges/"""
    shutil.rmtree('gh_pages/badges')
    shutil.copytree('badges', 'gh_pages/badges')

def publish():
    """Regenerate and publish to GitHub Pages"""
    x = local('git branch', capture=True)
    if '* master' not in x.stdout:
        print("ERROR: publish must be run from the master branch")
        raise SystemExit(1)
    x = local('git status', capture=True)
    if (('Your branch is up-to-date with' not in x.stdout and
                'HEAD detached at' not in x.stdout) or
                'nothing to commit' not in x.stdout):
        print("ERROR: Your local git clone is dirty or not pushed to origin.")
        raise SystemExit(1)
    local("ghp-import gh_pages")
    print("Changes pushed into gh-pages branch; please verify that branch and then push it to origin to deploy.")
